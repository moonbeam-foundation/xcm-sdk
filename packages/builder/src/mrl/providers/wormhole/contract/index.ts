import { Batch } from './Batch';
import { Gmp } from './Gmp';
import { TokenBridge } from './TokenBridge';
import { TokenBridgeRelayer } from './TokenBridgeRelayer';

export function contract() {
  return { Batch, Gmp, TokenBridge, TokenBridgeRelayer };
}
