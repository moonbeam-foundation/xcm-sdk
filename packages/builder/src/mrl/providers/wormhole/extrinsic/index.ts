import { ethereumXcm } from './ethereumXcm/ethereumXcm';
import { polkadotXcm } from './polkadotXcm/polkadotXcm';

export function extrinsic() {
  return { ethereumXcm, polkadotXcm };
}
