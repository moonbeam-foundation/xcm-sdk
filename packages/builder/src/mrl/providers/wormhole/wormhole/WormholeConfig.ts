import type { Network, Wormhole } from '@wormhole-foundation/sdk-connect';

export type WormholeTransferFunctions = 'tokenTransfer';
export type WormholeFunctionArgs = Parameters<
  Wormhole<Network>[WormholeTransferFunctions]
>;

export interface WormholeConfigConstructorParams {
  args: WormholeFunctionArgs;
  func: WormholeTransferFunctions;
}

export class WormholeConfig {
  readonly args: WormholeFunctionArgs;

  readonly func: WormholeTransferFunctions;

  static is(obj: unknown): obj is WormholeConfig {
    return obj instanceof WormholeConfig;
  }

  constructor({ args, func }: WormholeConfigConstructorParams) {
    this.args = args;
    this.func = func;
  }
}
