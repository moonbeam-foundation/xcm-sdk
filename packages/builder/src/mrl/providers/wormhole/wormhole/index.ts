export * from './wormhole';
export * from './WormholeConfig';
export * from './wormholeFactory';
