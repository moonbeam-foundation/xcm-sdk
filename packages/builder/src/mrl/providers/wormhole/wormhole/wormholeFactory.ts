import type { AnyChain } from '@moonbeam-network/xcm-types';
import { Wormhole } from '@wormhole-foundation/sdk-connect';
import { EvmPlatform } from '@wormhole-foundation/sdk-evm';

export function wormholeFactory(chain: AnyChain) {
  return new Wormhole(chain.isTestChain ? 'Testnet' : 'Mainnet', [EvmPlatform]);
}
