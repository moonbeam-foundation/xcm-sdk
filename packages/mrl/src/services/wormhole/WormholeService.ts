import {
  type WormholeConfig,
  wormholeFactory,
} from '@moonbeam-network/xcm-builder';
import '@wormhole-foundation/sdk-evm-tokenbridge';
import type { EvmSigner } from '@moonbeam-network/xcm-sdk';
import type { EvmChain, EvmParachain } from '@moonbeam-network/xcm-types';
import {
  TokenTransfer,
  type TransferQuote,
  type Wormhole,
} from '@wormhole-foundation/sdk-connect';
import { WormholeWagmiSigner } from './WormholeWagmiSigner';

export class WormholeService {
  readonly chain: EvmChain | EvmParachain;

  readonly #wh: Wormhole<'Mainnet' | 'Testnet'>;

  static create(chain: EvmChain | EvmParachain): WormholeService {
    return new WormholeService(chain);
  }

  constructor(chain: EvmChain | EvmParachain) {
    this.chain = chain;
    this.#wh = wormholeFactory(chain);
  }

  async getFee(transfer: WormholeConfig): Promise<TransferQuote | undefined> {
    const amount = transfer.args[1];
    if (amount === 0n) {
      return undefined;
    }

    const xfer = await this.#wh[transfer.func](...transfer.args);

    return TokenTransfer.quoteTransfer(
      this.#wh,
      xfer.fromChain,
      xfer.toChain,
      xfer.transfer,
    );
  }

  async transfer(
    signer: EvmSigner,
    transfer: WormholeConfig,
  ): Promise<string[]> {
    const xfer = await this.#wh[transfer.func](...transfer.args);

    return xfer.initiateTransfer(new WormholeWagmiSigner(this.chain, signer));
  }

  async getVaa(txId: string) {
    return await TokenTransfer.getTransferVaa(this.#wh, txId);
  }

  async getVaaBytes(vaa: TokenTransfer.VAA) {
    return (
      (await this.#wh.getVaaBytes({
        chain: vaa.emitterChain,
        emitter: vaa.emitterAddress,
        sequence: vaa.sequence,
      })) || undefined
    );
  }

  async getTokenTransfer(vaa: TokenTransfer.VAA, txId: string) {
    return await TokenTransfer.from(this.#wh, {
      chain: vaa.emitterChain,
      txid: txId,
    });
  }

  async isComplete(vaa: TokenTransfer.VAA, tokenTransfer: TokenTransfer) {
    const isComplete = await TokenTransfer.isTransferComplete(
      tokenTransfer.toChain,
      vaa,
    );
    return isComplete;
  }

  async completeTransfer(
    tokenTransfer: TokenTransfer,
    signer: EvmSigner,
  ): Promise<string> {
    const txIds = await tokenTransfer.completeTransfer(
      new WormholeWagmiSigner(this.chain, signer),
    );

    return txIds[0];
  }
}
