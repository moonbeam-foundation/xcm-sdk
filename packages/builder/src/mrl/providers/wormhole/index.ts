import { contract } from './contract';
import { extrinsic } from './extrinsic';
import { wormhole as whBuilder } from './wormhole';

export * from './extrinsic/ethereumXcm/BatchContractAbi';
export * from './wormhole/WormholeConfig';
export * from './wormhole/wormholeFactory';

export function wormhole() {
  return {
    contract,
    extrinsic,
    wormhole: whBuilder,
  };
}
