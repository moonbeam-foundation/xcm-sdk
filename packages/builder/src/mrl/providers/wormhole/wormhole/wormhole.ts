import { EvmChain, EvmParachain, Parachain } from '@moonbeam-network/xcm-types';
import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import { evmToAddress } from '@polkadot/util-crypto/address';
import { Wormhole } from '@wormhole-foundation/sdk-connect';
import { getExtrinsicAccount } from '../../../../extrinsic/ExtrinsicBuilder.utils';
import type {
  MrlBuilderParams,
  MrlConfigBuilder,
} from '../../../MrlBuilder.interfaces';
import { GMP_CONTRACT_ADDRESS } from '../contract/Gmp';
import { WormholeConfig } from './WormholeConfig';
import { wormholeFactory } from './wormholeFactory';

export function wormhole() {
  return {
    tokenTransfer: (): MrlConfigBuilder => ({
      build: ({
        asset,
        destination,
        destinationAddress,
        isAutomatic,
        moonApi,
        moonChain,
        source,
        sourceAddress,
      }): WormholeConfig => {
        const isSourceParachain = Parachain.is(source);
        const isDestinationMoonChain = destination.isEqual(moonChain);
        const isDestinationEvmChain = EvmChain.is(destination);
        const isNativeAsset = asset.isSame(
          isDestinationEvmChain ? moonChain.nativeAsset : source.nativeAsset,
        );
        const tokenAddress = isNativeAsset
          ? 'native'
          : isDestinationEvmChain
            ? moonChain.getChainAsset(asset).address
            : asset.address;

        const { address20: computedOriginAccount } =
          getMultilocationDerivedAddresses({
            address: sourceAddress,
            paraId: isSourceParachain ? source.parachainId : undefined,
            isParents: true,
          });

        if (!tokenAddress) {
          throw new Error(`Asset ${asset.key} has no address`);
        }

        const wh = wormholeFactory(source);
        const whSource = isDestinationEvmChain
          ? wh.getChain(moonChain.getWormholeName())
          : wh.getChain(source.getWormholeName());
        const whDestination = isDestinationEvmChain
          ? wh.getChain(destination.getWormholeName())
          : wh.getChain(moonChain.getWormholeName());
        const whAsset = Wormhole.tokenId(whSource.chain, tokenAddress);
        const whSourceAddress = Wormhole.chainAddress(
          whSource.chain,
          isDestinationEvmChain ? computedOriginAccount : sourceAddress,
        );
        const whDestinationAddress = Wormhole.chainAddress(
          whDestination.chain,
          isDestinationMoonChain || isDestinationEvmChain
            ? destinationAddress
            : GMP_CONTRACT_ADDRESS,
        );

        return new WormholeConfig({
          args: [
            whAsset,
            asset.amount,
            whSourceAddress,
            whDestinationAddress,
            isAutomatic,
            isDestinationMoonChain || isDestinationEvmChain
              ? undefined
              : getPayload({ destination, destinationAddress, moonApi }),
          ],
          func: 'tokenTransfer',
        });
      },
    }),
  };
}

/*
 * Extrinsic to GMP precompile
 * https://docs.moonbeam.network/builders/ethereum/precompiles/interoperability/gmp/
 */
export function getPayload({
  moonApi,
  destination,
  destinationAddress,
}: Pick<MrlBuilderParams, 'destination' | 'destinationAddress' | 'moonApi'>):
  | Uint8Array
  | undefined {
  if (!EvmParachain.isAnyParachain(destination)) {
    throw new Error(
      `Destination ${destination.name} is not a Parachain or EvmParachain`,
    );
  }

  // TODO: This is workaround for Peaq EVM. Can we improve it so we don't need to have hardcoded values?
  const isPeaqEvm =
    destination.key === 'peaq-evm-Alphanet' || destination.key === 'peaq-evm';

  const multilocation = moonApi.createType('XcmVersionedLocation', {
    V3: {
      parents: 1,
      interior: {
        X2: [
          {
            Parachain: destination.parachainId,
          },
          getExtrinsicAccount(
            isPeaqEvm ? evmToAddress(destinationAddress) : destinationAddress,
          ),
        ],
      },
    },
  });
  const action = moonApi.createType('XcmRoutingUserAction', {
    destination: multilocation,
  });
  const versioned = moonApi.createType('VersionedUserAction', {
    V1: action,
  });

  return versioned.toU8a();
}
