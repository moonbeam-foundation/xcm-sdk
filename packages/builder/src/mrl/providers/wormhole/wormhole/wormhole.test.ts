import { describe, expect, it } from 'vitest';

import { getPolkadotApi } from '@moonbeam-network/xcm-utils';
import {
  wormholeConfigBuilderParams,
  wormholeToMoonchainConfigBuilderParams,
} from '../../../../../fixtures';
import { wormhole } from './wormhole';

describe('wormhole', async () => {
  const moonApi = await getPolkadotApi(
    'wss://wss.api.moonbase.moonbeam.network',
  );

  describe('tokenTransfer with isAutomatic=true', () => {
    const transfer = wormhole().tokenTransfer();

    it('should be correct config', () => {
      expect(
        transfer.build({
          ...wormholeConfigBuilderParams,
          moonApi,
          isAutomatic: true,
        }),
      ).toMatchSnapshot();
    });
  });

  describe('tokenTransfer with isAutomatic=false', () => {
    const transfer = wormhole().tokenTransfer();

    it('should be correct config to moon chain', () => {
      expect(
        transfer.build({
          ...wormholeToMoonchainConfigBuilderParams,
          moonApi,
          isAutomatic: false,
        }),
      ).toMatchSnapshot();
    });
  });
});
