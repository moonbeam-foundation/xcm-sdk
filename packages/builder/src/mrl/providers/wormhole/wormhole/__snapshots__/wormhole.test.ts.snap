// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wormhole > tokenTransfer with isAutomatic=false > should be correct config to moon chain 1`] = `
WormholeConfig {
  "args": [
    {
      "address": "native",
      "chain": "Fantom",
    },
    99000000000n,
    {
      "address": EvmAddress {
        "address": "0xeF46c7649270C912704fB09B75097f6E32208b85",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    {
      "address": EvmAddress {
        "address": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
        "type": "Native",
      },
      "chain": "Moonbeam",
    },
    false,
    undefined,
  ],
  "func": "tokenTransfer",
}
`;

exports[`wormhole > tokenTransfer with isAutomatic=true > should be correct config 1`] = `
WormholeConfig {
  "args": [
    {
      "address": "native",
      "chain": "Fantom",
    },
    99000000000n,
    {
      "address": EvmAddress {
        "address": "0xeF46c7649270C912704fB09B75097f6E32208b85",
        "type": "Native",
      },
      "chain": "Fantom",
    },
    {
      "address": EvmAddress {
        "address": "0x0000000000000000000000000000000000000816",
        "type": "Native",
      },
      "chain": "Moonbeam",
    },
    true,
    Uint8Array [
      0,
      3,
      1,
      2,
      0,
      165,
      15,
      1,
      0,
      212,
      53,
      147,
      199,
      21,
      253,
      211,
      28,
      97,
      20,
      26,
      189,
      4,
      169,
      159,
      214,
      130,
      44,
      133,
      88,
      133,
      76,
      205,
      227,
      154,
      86,
      132,
      231,
      165,
      109,
      162,
      125,
    ],
  ],
  "func": "tokenTransfer",
}
`;
